src/compiler/scala/tools/nsc/classpath/DirectoryClassPath.scala

```diff
@@ -94,28 +111,32 @@ trait JFileDirectoryLookup[FileEntryType <: ClassRepresentation] extends Directo
     if (packageDir.exists && packageDir.isDirectory && packageDir.canRead) Some(packageDir)
     else None
   }
-  protected def listChildren(dir: File, filter: Option[File => Boolean]): Array[File] = {
-    val listing = filter match {
-      case Some(f) => dir.listFiles(mkFileFilter(f))
-      case None => dir.listFiles()
-    }
 
-    // Sort by file name for stable order of directory .class entries in package scope.
-    // This gives stable results ordering of base type sequences for unrelated classes
-    // with the same base type depth.
-    //
-    // Notably, this will stably infer`Product with Serializable`
-    // as the type of `case class C(); case class D(); List(C(), D()).head`, rather than the opposite order.
-    // On Mac, the HFS performs this sorting transparently, but on Linux the order is unspecified.
-    //
-    // Note this behaviour can be enabled in javac with `javac -XDsortfiles`, but that's only
-    // intended to improve determinism of the compiler for compiler hackers.
-    util.Arrays.sort(listing, (o1: File, o2: File) => o1.getName.compareTo(o2.getName))
-    listing
+  override protected def listChildren(dir: File)(f: (File, BasicFileAttributes) => Boolean): (Array[File], Array[File]) = {
+    import java.nio.file._
+    val files = new collection.mutable.ArrayBuilder.ofRef[File]()
+    val dirs = new collection.mutable.ArrayBuilder.ofRef[File]()
+    val visitor = new SimpleFileVisitor[Path]() {
+      override def visitFile(childFile: Path, attrs: attribute.BasicFileAttributes): FileVisitResult = {
+        if (dir != childFile && f(childFile.toFile, attrs)) {
+          if (attrs.isDirectory) dirs += childFile.toFile
+          else files += childFile.toFile
+        }
+        FileVisitResult.CONTINUE
```

Comment:
@mkeskells You mentioned that you benchmarked a few different ways do do this in similar work you did over in SBT. Could you provide some pointers?

My research led me to believe that this approach minimized the number of underlying file system calls by grabbing all the file attributes in one go, rather than before when subsequent calls to `isDirectory` and `exists` would call to the filesystem again.